source/base/tensor_product_polynomials.cc

```diff
@@ -277,159 +277,110 @@ compute (const Point<dim>            &p,
-  if (n_polynomials > 20)
-    {
-      large_array.resize(n_polynomials);
-      v = &large_array[0];
-    }
+  boost::container::small_vector<std::array<std::array<double,5>,dim>, 20> values_1d(n_polynomials);
+  if (n_values_and_derivatives == 1)
+    for (unsigned int i=0; i<n_polynomials; ++i)
+      for (unsigned int d=0; d<dim; ++d)
+        values_1d[i][d][0] = polynomials[i].value(p(d));
   else
-    v = &small_array[0];
-
-  for (unsigned int i=0; i<n_polynomials; ++i)
-    for (unsigned int d=0; d<dim; ++d)
-      polynomials[i].value(p(d), n_values_and_derivatives, &v[i][d][0]);
-
-  for (unsigned int i=0; i<n_tensor_pols; ++i)
-    {
-      // first get the
-      // one-dimensional indices of
-      // this particular tensor
-      // product polynomial
-      unsigned int indices[dim];
-      compute_index (i, indices);
-
-      if (update_values)
-        {
-          values[i] = 1;
-          for (unsigned int x=0; x<dim; ++x)
-            values[i] *= v[indices[x]][x][0];
-        }
-
-      if (update_grads)
-        for (unsigned int d=0; d<dim; ++d)
+    for (unsigned int i=0; i<n_polynomials; ++i)
+      for (unsigned d=0; d<dim; ++d)
+        polynomials[i].value(p(d), n_values_and_derivatives, &values_1d[i][d][0]);
+
+  unsigned int indices[3];
+  unsigned int ind=0;
+  for (indices[2]=0; indices[2]<(dim>2?n_polynomials:1); ++indices[2])
+    for (indices[1]=0; indices[1]<(dim>1?n_polynomials:1); ++indices[1])
+      if (n_values_and_derivatives == 1)
+        for (indices[0]=0; indices[0]<n_polynomials; ++indices[0], ++ind)
           {
-            grads[i][d] = 1.;
-            for (unsigned int x=0; x<dim; ++x)
-              grads[i][d] *= v[indices[x]][x][d==x];
+            double value = values_1d[indices[0]][0][0];
+            for (unsigned int d=1; d<dim; ++d)
+              value *= values_1d[indices[d]][d][0];
+            values[index_map_inverse[ind]] = value;
           }
+      else
+        for (indices[0]=0; indices[0]<n_polynomials; ++indices[0], ++ind)
+          {
+            unsigned int i = index_map_inverse[ind];
 
-      if (update_grad_grads)
-        for (unsigned int d1=0; d1<dim; ++d1)
-          for (unsigned int d2=0; d2<dim; ++d2)
-            {
-              grad_grads[i][d1][d2] = 1.;
-              for (unsigned int x=0; x<dim; ++x)
-                {
-                  unsigned int derivative=0;
-                  if (d1==x) ++derivative;
-                  if (d2==x) ++derivative;
+            if (update_values)
+              {
+                double value = values_1d[indices[0]][0][0];
+                for (unsigned int x=1; x<dim; ++x)
+                  value *= values_1d[indices[x]][x][0];
+                values[i] = value;
+              }
 
-                  grad_grads[i][d1][d2]
-                  *= v[indices[x]][x][derivative];
+            if (update_grads)
+              for (unsigned int d=0; d<dim; ++d)
+                {
+                  double grad = 1.;
```

Comment:
Why do you multiply into a temporary here but not in all the other cases below? I assume the compiler doesn't care?

Reply:
I thought so too, but it turned out that the compiler then must write the content of `grads[i][d]` to memory after each step because it cannot prove that the array does not alias with values_1d, so that could potentially be changed in that loop.

Reply:
For the higher derivatives I simply didn't care. We should do it uniformly so I will apply this change to all cases below.